dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/SignalN/GeneratorPVDiagramPQ.mo

```diff
@@ -0,0 +1,80 @@
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, you can obtain one at http://mozilla.org/MPL/2.0/.
+* SPDX-License-Identifier: MPL-2.0
+*
+* This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+model GeneratorPVDiagramPQ "Model for generator PV based on SignalN for the frequency handling with an N points PQ diagram."
+
+  import Dynawo.Connectors;
+  import Dynawo.Electrical.Machines;
+  import Modelica;
+  extends Machines.BaseClasses.BaseGeneratorSimplified;
+  extends AdditionalIcons.Machine;
+
+    parameter Types.ActivePowerPu PMinPu "Minimum active power in p.u (base SnRef)";
+    parameter Types.ActivePowerPu PMaxPu "Maximum active power in p.u (base SnRef)";
+    parameter Types.ReactivePowerPu QMin0Pu "Start value of the minimum reactive power in p.u (base SnRef)";
+    parameter Types.ReactivePowerPu QMax0Pu "Start value of the maximum reactive power in p.u (base SnRef)";
+    parameter Types.PerUnit KGover "Mechanical power sensitivity to frequency";
+    parameter Types.ActivePower PNom "Nominal power in MW";
+    final parameter Real Alpha = PNom * KGover "Participation of the considered generator in the frequency regulation";
+    parameter Types.Time tFilter "Filter time constant to update QMin/QMax";
+    parameter String QMinTableName "Name of the table in the text file to get QMinPu from PGenPu";
+    parameter String QMaxTableName "Name of the table in the text file to get QMaxPu from PGenPu";
+    parameter String QMinTableFile "Text file that contains the table to get QMinPu from PGenPu";
+    parameter String QMaxTableFile "Text file that contains the table to get QMaxPu from PGenPu";
+
+    Connectors.ImPin N "Signal to change the active power reference setpoint of all the generators in the system in p.u (base SnRef)";
+    Connectors.ZPin alpha "Participation of the considered generator in the frequency regulation";
+    Connectors.ZPin alphaSum "Sum of all the participations of all generators in the frequency regulation";
+    Connectors.ZPin URefPu (value(start = U0Pu)) "Voltage regulation set point in p.u (base UNom)";
+    Modelica.Blocks.Tables.CombiTable1D tableQMin(tableOnFile = true, tableName = QMinTableName, fileName = QMinTableFile) "Table to get QMinPu from PGenPu";
+    Modelica.Blocks.Tables.CombiTable1D tableQMax(tableOnFile = true, tableName = QMaxTableName, fileName = QMaxTableFile) "Table to get QMaxPu from PGenPu";
+    Types.ReactivePowerPu QMinPu(start = QMin0Pu) "Minimum reactive power in p.u (base SnRef)";
+    Types.ReactivePowerPu QMaxPu(start = QMax0Pu) "Maximum reactive power in p.u (base SnRef)";
+
+  protected
+
+    Types.ActivePowerPu PGenRawPu (start = PGen0Pu) "Active power generation without taking limits into account in p.u (base SnRef) (generator convention)";
+
+  equation
+
+    PGenPu = tableQMin.u[1];
+    tFilter * der(QMinPu) + QMinPu = tableQMin.y[1];
```

Comment:
I must admit that I don't like the use of QMin0Pu with the 0 in it (same in INIT model). Could you explain why did you need the variables QMinPu and QMaxPu with a dynamic evolution? Otherwise we could have had only QMinPu, QMaxPu as parameters and use tableQMin.y as final limit and same for tableQMax.y. Or would you have another suggestion/idea?

Reply:
Well if I put QMinPu and QMaxPu as parameters, and use tableQMin.y[1] and tableQMax.y[1] in the model as the actual QMinPu and QMaxPU (they change because they depend on PGenPu), I think it is quite confusing. My opinion is that we leave the impression that QMinPu and QMaxPu are fixed while they are in fact variables that depend on PGenPu. With the 0 I just give the Init value of QMinPu and QMaxPu, and they are then calculated in the model for each time step.
I could simply write QMinPu = tableQMin.y[1] (without the first order filter) but I have numerical issues. This tFilter parameter will use a default value and won't be chosen by the user. What do you think ?

Reply:
Ok then keep the dynamic ;)
But If the limits depend on PGenPu, how can you set the QMin0Pu and QMax0Pu to the right value without going through the tables?

Reply:
Good question! In fact they only need to respect QMin0Pu<Q0Pu<QMax0Pu, so I choose the largest value of the table (then it is replaced with the actual value at the first time step (maybe global init)). If I don't put an init value, it chooses 0, so I have QMinPu and QMaxPu that start at 0 with Q<>0, and it fails.

Reply:
Ok, in that case, could we have something like:
Types.ReactivePowerPu QMinPu(start = tableQMin[X])
in order not to have QMin0Pu in the par file (as it is not a real value...)?

Reply:
I'll try something like that and let you know

Reply:
I doesn't seem to work this way...
It could work if the table was defined in the .mo but it is read in the .txt file so I cannot easily access it through Modelica.

Reply:
Ok thanks for trying